openml/extensions/extension_interface.py

```diff
@@ -0,0 +1,291 @@
+from abc import ABC, abstractmethod
+from collections import OrderedDict  # noqa: F401
+from typing import Any, Dict, List, Optional, Tuple, TYPE_CHECKING
+
+
+if TYPE_CHECKING:
+    from openml.flows import OpenMLFlow
+    from openml.tasks.task import OpenMLTask
+    from openml.runs.trace import OpenMLRunTrace, OpenMLTraceIteration
+
+
+class Extension(ABC):
+
+    """Defines the interface to connect machine learning libraries to OpenML-Python.
+
+    See ``openml.extension.sklearn.extension`` for an implementation to bootstrap from.
+    """
+
+    ################################################################################################
+    # General setup
+
+    @staticmethod
+    @abstractmethod
+    def can_handle_flow(flow: 'OpenMLFlow') -> bool:
+        """Check whether a given flow can be handled by this extension.
+
+        This is typically done by parsing the ``external_version`` field.
+
+        Parameters
+        ----------
+        flow : OpenMLFlow
+
+        Returns
+        -------
+        bool
+        """
+        pass
+
+    @staticmethod
+    @abstractmethod
+    def can_handle_model(model: Any) -> bool:
+        """Check whether a model flow can be handled by this extension.
+
+        This is typically done by checking the type of the model, or the package it belongs to.
+
+        Parameters
+        ----------
+        model : Any
+
+        Returns
+        -------
+        bool
+        """
+        pass
```

Comment:
rather than a pass statement, should we throw a `NotImplementedError`? That is way more explicit and gives better feedback to users when they accidentally use this abstract extension.

Reply:
It's actually possible to remove the `pass`, too. Making sure that the abstract method is not used is done by the decorator which will raise an exception if you try to create a subclass which does not implement the method.